Ignore results annotated with '# noqa'

RJ722 opened the pull request:
Hey, 

## Description

This PR isn't production-ready by any means, but I just wanted to have an idea about the complexity it would add to Vulture to support `# noqa` comments. I'd be happy to add tests and documentation, once we all settle if it would be a good idea to carry this forward.

Another change with the introduction of `# noqa` comments would be the introduction of issue codes (it's possible without these as well, but would hurt specificity). For now, I've just added `V001`, `V002`, ... - the nomenclature for these issue codes is another thing to think about.

## Related Issue
<!--- Ideally, new features and changes are discussed in an issue first. -->
<!--- If there is a corresponding issue, please link to it here: -->

#160, #182, #190 

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation in the README file.
- [x] I have updated the documentation accordingly.
- [x] I have added an entry in CHANGELOG.md.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

RJ722 commented on vulture/core.py:
```diff
@@ -46,6 +47,39 @@
+    # We want a comma-separated list of errors
+    # https://regex101.com/r/4XUuax/3 full explenation of the regex
+    r"#( )?noqa(?::[\s]?(?P<codes>([A-Z]+[0-9]+(?:[,\s]+)?)+))?",
+    re.IGNORECASE,
```
This expression was taken from https://gitlab.com/pycqa/flake8//blob/0ac337608b3d88079aa91bdcffed7b4df6fd28a8/src/flake8/defaults.py#L26 with an added case of `#noqa` (without a space b/w `#` and `noqa`) -- explanation URL updated.

jendrikseipp replied:
I think it makes sense to follow exactly what flake8 uses.

RJ722 replied:
Okay. Changed it back.

jendrikseipp commented on vulture/core.py:
```diff
@@ -129,6 +164,7 @@ def __init__(
         self.last_lineno = last_lineno
         self.message = message or "unused {typ} '{name}'".format(**locals())
         self.confidence = confidence
+        self.issue_code = ISSUE_CODE_MAP[typ]
```
No need to store this here if it's easily deducible from the `typ`.

jendrikseipp commented on vulture/core.py:
```diff
@@ -201,14 +237,27 @@ def get_set(typ):
         self.found_dead_code_or_error = False
 
+    def add_noqa_match(self, code, lineno):
+        self.noqa_matches[code].append(lineno)
```
No separate method needed.

jendrikseipp commented on vulture/core.py:
```diff
@@ -46,6 +47,39 @@
+)
+
+
+def _find_noqa(line):
```
No separate method needed.

jendrikseipp commented on vulture/core.py:
```diff
@@ -201,14 +237,27 @@ def get_set(typ):
+        self.noqa_matches[code].append(lineno)
+
+    def parse_noqa(self, code):
+        for index, line in enumerate(code):
```
enumerate(code, start=1)

RJ722 replied:
thanks!

jendrikseipp commented on vulture/core.py:
```diff
@@ -201,14 +237,27 @@ def get_set(typ):
+    def parse_noqa(self, code):
+        for index, line in enumerate(code):
+            match = _find_noqa(line)
+            if match is not None:
```
`if match`

jendrikseipp commented on vulture/core.py:
```diff
@@ -201,14 +237,27 @@ def get_set(typ):
+        for index, line in enumerate(code):
+            match = _find_noqa(line)
+            if match is not None:
+                code = match.groupdict()["codes"] or "all"
```
This needs to handle the case when multiple codes are given. (split by commas and strip whitespace, I guess.)

RJ722 replied:
Done!

jendrikseipp commented on vulture/core.py:
```diff
@@ -282,6 +331,16 @@ def get_unused_code(self, min_confidence=0, sort_by_size=False):
+                if obj.issue_code in self.noqa_matches
+                else "all"
+            )
+            if code in self.noqa_matches:
```
I think this line (the if-test) can be removed.

jendrikseipp commented on vulture/core.py:
```diff
@@ -201,14 +237,27 @@ def get_set(typ):
 
         self.ignore_names = ignore_names or []
         self.ignore_decorators = ignore_decorators or []
+        self.noqa_matches = defaultdict(list)
```
This should be a set instead of a list.

jendrikseipp commented on vulture/core.py:
```diff
@@ -301,8 +360,14 @@ def by_size(item):
 
+        results_without_noqa = [
+            obj for obj in unused_code if not _has_noqa(obj)
+        ]
```
No need to create a filtered list. Simply filter objects below.

RJ722 replied:
Yeah, that improves readability. thanks!

jendrikseipp commented on vulture/core.py:
```diff
@@ -282,6 +331,16 @@ def get_unused_code(self, min_confidence=0, sort_by_size=False):
+                else "all"
+            )
+            if code in self.noqa_matches:
+                return obj.first_lineno in self.noqa_matches[code]
```
I think the whole function can be reduced to `return obj.first_lineno in self.noqa_matches["all"] or obj.first_lineno in self.noqa_matches[ISSUE_CODE_MAP[obj.typ]]`

RJ722 replied:
I ended up with:

```py
return obj.first_lineno in (
    self.noqa_matches[ISSUE_CODE_MAP[obj.typ]]
    or self.noqa_matches["all"]
)
```

RJ722 commented:
@jendrikseipp should `ISSUE_CODE_MAP` be a class variable for `Vulture`?

I'm asking because my guess is that most folks would only import Vulture, if using the API (`from vulture import Vulture`).

jendrikseipp commented on vulture/core.py:
```diff
@@ -46,6 +47,34 @@
 if sys.version_info < (3, 4):
     IGNORED_VARIABLE_NAMES |= {"True", "False"}
 
+ISSUE_CODE_MAP = {
```
Let's call this dict ERROR_CODES. I think leaving this as a global variable is fine.

jendrikseipp commented on vulture/core.py:
```diff
@@ -282,6 +325,13 @@ def get_unused_code(self, min_confidence=0, sort_by_size=False):
+            return obj.first_lineno in (
+                self.noqa_matches[ISSUE_CODE_MAP[obj.typ]]
+                or self.noqa_matches["all"]
+            )
```
This will fail if, e.g., "noqa V001" is given for line 10, "noqa" is given for line 5 and the object appears on line 5.

RJ722 replied:
Oh yes, nice catch. We need a union of two sets here.

jendrikseipp commented on vulture/core.py:
```diff
@@ -282,6 +325,13 @@ def get_unused_code(self, min_confidence=0, sort_by_size=False):
         Return ordered list of unused Item objects.
         """
+
+        def _has_noqa(obj):
```
No reason for leading underscore in nested function.

jendrikseipp commented on vulture/core.py:
```diff
@@ -46,6 +47,34 @@
+    # We do not want to capture the ``: `` that follows ``noqa``
+    # We do not care about the casing of ``noqa``
+    # We want a comma-separated list of errors
+    # https://regex101.com/r/4XUuax/2 full explanation of the regex
```
No need to copy the whole comment. The first three lines and a well-formed example are enough.

jendrikseipp commented on vulture/core.py:
```diff
@@ -201,14 +230,28 @@ def get_set(typ):
+                codes = [
+                    c.strip()
+                    for c in (match.groupdict()["codes"] or "all").split(",")
+                    # if no code is specified, assign it to `all`
```
Comment above code.

jendrikseipp commented on vulture/core.py:
```diff
@@ -301,8 +351,11 @@ def by_size(item):
         )
 
+        # breakpoint()
         confidently_unused = [
```
Rename to "reported_unused".

RJ722 commented:
Another thing I was pondering about. We don't start ignoring `noqa` until `get_unused_code` is called. What if the user bypasses it, e.g.:

```py
from vulture import Vulture

v = Vulture()
v.scan('import this  # noqa: V004')
print(v.unused_imports)
```

Should we move the `has_noqa` logic to `_define` method?

jendrikseipp commented:
> Should we move the `has_noqa` logic to `_define` method?

Good idea! This aligns better with the existing code for ignoring things.

RJ722 commented:
TODO:

- [ ] One test is failing because Python3.8 onwards, the first line for property node is the same line at which the function starts.
- [ ] Write documentation for these changes -- @jendrikseipp can you take this on, please? 😅

RJ722 commented on tests/test_noqa.py:
```diff
@@ -0,0 +1,205 @@
+    v.scan(
+        """\
+import abc  # noqa: V012, V😎12
+import problems  # noqa: V03🙃1
```
This test is the culprit behind tests failing for Python2.7. Should I remove them?

jendrikseipp replied:
Yes, I think it's easiest to remove the emojis.

jendrikseipp commented:
Could you also add a test that verifies that we fixed the "or vs. union" bug?

RJ722 commented:
> Could you also add a test that verifies that we fixed the "or vs. union" bug?

The way I've added the tests is basically ensuring that an example of the same issue code is  covered in both categories - `all` and `V00X` (on different lines - both before and after). If there'd have been a problem with "or v/s union", we'd have seen it by now.

RJ722 commented:
But, now that I come think of it, I haven't added isolated cases - just `V00X`, or `# noqa` alone. Adding them.

RJ722 commented on vulture/core.py:
```diff
@@ -457,26 +495,41 @@ def _define(
+            if typ == "property" and sys.version_info >= (3, 8):
+                # python3.8 onwards, lineno for property is the same as that of
+                # the decorated method.
+                lineno = lineno - 1
```
Also, I think, it rather makes sense to ignore this case:

```py
class Foo:
    @property
    def bar():  # noqa
        pass
```

than adding `# noqa` infront of `@property`?

jendrikseipp replied:
I think adding the noqa on the same line as bar() is the better solution, yes. We should change this at a lower level as well: in all Python versions, we should report the line containing bar() as the line number for the property. This should be a separate PR though. However, this might be tricky in case there are multiple @property decorators. Could you try this out, please?

RJ722 commented on tests/test_noqa.py:
```diff
@@ -0,0 +1,245 @@
+
+@pytest.mark.parametrize(
+    "line",
+    [("# noqa: 123V"), ("# noqa explaination: V012"), ("# noqa: ,V001"),],
```
Remove trailing comma.

jendrikseipp commented on tests/test_noqa.py:
```diff
@@ -0,0 +1,245 @@
+def test_noqa_regex_present(line, codes):
+    match = NOQA_REGEXP.search(line)
+    parsed = [
+        c.strip() for c in (match.groupdict()["codes"] or "all").split(",")
```
Can you factor this out into a function in core.py?

jendrikseipp commented on tests/test_report.py:
```diff
@@ -39,13 +39,13 @@ def test_report(code, expected, make_whitelist=False):
+<V001> {filename}:8: unused attribute 'foobar' (60% confidence)
+<V007> {filename}:9: unused variable 'foobar' (60% confidence)
+<V006> {filename}:11: unreachable code after 'return' (100% confidence)
+<V005> {filename}:13: unused property 'myprop' (60% confidence)
```
I think we should mimic the flake8 format more closely (test.py:2:5: F841 local variable 'foo' is assigned to but never used), so for example:

{filename}:1: V004 unused import 'foo' (90% confidence)

jendrikseipp commented on tests/test_noqa.py:
```diff
@@ -0,0 +1,245 @@
+    def serve(self, quantity=50):  # noqa: V003
+        self.quantity_served = quantity  # noqa: V001
+        return
+        self.pour()  # noqa: V006
```
Shouldn't this be V007?

jendrikseipp commented on vulture/core.py:
```diff
@@ -457,26 +495,41 @@ def _define(
-            name, self.ignore_names
-        ):
+        first_lineno = first_node.lineno
+        if ignored(name) or has_noqa(first_lineno, typ):
```
`has_noqa` can become part of `ignored`

jendrikseipp commented on vulture/core.py:
```diff
@@ -46,6 +47,28 @@
+    "property": "V005",
+    "unreachable_code": "V006",
+    "variable": "V007",
+}
```
I think we should follow pycodestyle and start with error code V101 instead of V001. Since unreachable_code is somewhat different from unused code, I'd give it number V201 (and unused variables V107). (I see now that you changed the order.)

RJ722 replied:
If we start with V101, and don't assign unreachable_code in that series, unused_variables would be V106, no?

jendrikseipp replied:
That's right :-)

jendrikseipp commented on tests/test_noqa.py:
```diff
@@ -0,0 +1,274 @@
+    [("#noqa"), ("##noqa"), ("# n o q a"), ("#NOQA"), ("# Hello, noqa")],
+)
+def test_noqa_regex_not_present(line):
+    assert bool(NOQA_REGEXP.search(line)) is False
```
Why not `assert not NOQA_REGEXP.search(line)`?

RJ722 replied:
I just didn't happen to think of it. ':D Changing it.

jendrikseipp commented on tests/test_noqa.py:
```diff
@@ -0,0 +1,274 @@
+    "line",
+    [
+        ("# noqa: 123V"),
+        ("# noqa explaination: V012"),
```
explaination -> explanation

jendrikseipp commented on tests/test_noqa.py:
```diff
@@ -0,0 +1,274 @@
+    check(v.unused_imports, [])
+    check(v.unused_props, [])
+    check(v.unreachable_code, [])
+    check(v.unused_vars, [])
```
Can you use exactly one noqa for each category?

RJ722 replied:
Do you mean, break it down into individual tests for all of them?

jendrikseipp replied:
No, I was just surprised to see 6 noqas but 7 checks. Adding the `year` variable here as well should resolve this.

RJ722 replied:
Aah, okay. I took care of this.

jendrikseipp commented on tests/test_noqa.py:
```diff
@@ -0,0 +1,274 @@
+
+class Cellar:  # noqa: V102
+    @property
+    def wine(self, grapes):  # noqa: V105, V106
```
Same here. let's add a separate unused variable inside of this method. The we can also remove the `pass`.

RJ722 replied:
I added the variable - `year`. Just making sure that you in fact, meant this. If yes, can we mark this as resolved?

jendrikseipp replied:
I'd drop the useless return and remove the V106 noqa from the line above, but then it's resolved, yes.

RJ722 replied:
Ahh, got it.

jendrikseipp commented on tests/test_noqa.py:
```diff
@@ -0,0 +1,274 @@
+
+# noqa
+def problems():  # noqa: V104
+    ''' They don't go away. :-)'''
```
Too pessimistic for my taste :-) I'd remove the comment.

jendrikseipp commented on tests/test_noqa.py:
```diff
@@ -0,0 +1,274 @@
+"""
+    )
+    check(v.unreachable_code, [])
+    check(v.unused_funcs, ["shave_sheep"])
```
Hehe, nice code :-)

jendrikseipp commented on tests/test_report.py:
```diff
@@ -39,13 +39,13 @@ def test_report(code, expected, make_whitelist=False):
+{filename}:8: V101 unused attribute 'foobar' (60% confidence)
+{filename}:9: V106 unused variable 'foobar' (60% confidence)
+{filename}:11: V201 unreachable code after 'return' (100% confidence)
+{filename}:13: V105 unused property 'myprop' (60% confidence)
```
Can you add to the changelog that the output format changes?

jendrikseipp commented on vulture/core.py:
```diff
@@ -461,26 +501,45 @@ def _define(
+                lineno = lineno + len(first_node.decorator_list)
+            return lineno in (
+                self.noqa_matches[ERROR_CODES[typ]].union(
+                    self.noqa_matches["all"]
```
I think it's faster and more readable to test the two sets separately instead of computing the union just for the test.

RJ722 replied:
That might result in the same or v/s union bug, I guess. Do you have something in mind?

jendrikseipp replied:
`return lineno in self.noqa_matches[ERROR_CODES[typ]] or lineno in self.noqa_matches["all"]`

RJ722 replied:
Yep, thanks!

jendrikseipp commented on vulture/core.py:
```diff
@@ -461,26 +501,45 @@ def _define(
+                # decorators on the method.
+                # python3.8 onwards, lineno for property is the same as that of
+                # the decorated method.
+                lineno = lineno + len(first_node.decorator_list)
```
I think this means that the line numbers we report for unused properties now differ from the line numbers that must have noqa on them. This should be fixed in a separate PR. Can you add a TODO to the TODO file, please?

RJ722 replied:
I raised #200 :D -- I would be happy to rebase this branch, once we get it merged.

jendrikseipp commented on vulture/core.py:
```diff
@@ -461,26 +501,45 @@ def _define(
+                # for python < 3.8, increment as many line numbers as there are
+                # decorators on the method.
+                # python3.8 onwards, lineno for property is the same as that of
+                # the decorated method.
```
Please start the first sentence with an uppercase "For" and the second with "From".

jendrikseipp commented on CHANGELOG.md:
```diff
@@ -1,6 +1,9 @@
 # Unreleased
 
 * Ignore unused import statements which occur in `__init__.py` (RJ722, #192).
+* Report issue codes in ouput (e.g., `code.py:1: V104 unused import ...`)
```
ouput -> output

jendrikseipp commented on vulture/core.py:
```diff
@@ -55,6 +78,13 @@ def _get_unused_items(defined_items, used_names):
 
 
+def _parse_error_codes(match):
+    # if no code is specified, assign it to `all`
```
Start with upper case and end with full stop. I#d also be a little more concrete: 

If no error code is specified, add the line to the "all" category.

jendrikseipp commented on vulture/core.py:
```diff
@@ -461,26 +501,44 @@ def _define(
+            ):
+                return True
+
+            # if name not ignored, check if annotated with "# noqa"
```
Rephrase: Check if the reported line is annotated with "# noqa".

RJ722 commented:
Now, that #200 has been merged (yay! \o/), let me rebase this branch and resolve conflicts.

jendrikseipp commented on CHANGELOG.md:
```diff
@@ -3,6 +3,9 @@
   (RJ722, #200)
+* Report issue codes in output (e.g., `code.py:1: V104 unused import ...`)
+  (RJ722, #195)
+* Support `# noqa` comments to suppress results on that line. (RJ722, #195).
```
Could you please draft the documentation for the README.md file? It's always easier to edit existing text than writing new text.

RJ722 replied:
Sure!

jendrikseipp commented on vulture/core.py:
```diff
@@ -205,14 +236,23 @@ def get_set(typ):
         self.code = []
         self.found_dead_code_or_error = False
 
+    def parse_noqa(self, code):
```
It's a bit strange that the method is passed `self.code`, but not `self.noqa_matches` even though both are instance attributes. I think we should pass both variables to the method and turn it into a function. Then we can collect all code related to noqa in a vulture/noqa.py file, i.e., parse_noqa(), _parse_error_codes() and NOQA_REGEXP.

jendrikseipp commented on vulture/core.py:
```diff
@@ -461,14 +501,25 @@ def _define(
+            return (
+                lineno in self.noqa_matches[ERROR_CODES[typ]]
+                or lineno in self.noqa_matches["all"]
+            )
```
This can become a function `noqa.ignore_line(noqa_matches, lineno, typ)` and then we can add it to the if-statement above.

jendrikseipp commented on vulture/core.py:
```diff
@@ -205,14 +236,23 @@ def get_set(typ):
 
         self.ignore_names = ignore_names or []
         self.ignore_decorators = ignore_decorators or []
+        self.noqa_matches = defaultdict(set)
```
Please rename noqa_matches to noqa_lines everywhere.

jendrikseipp commented:
Thanks for the additional changes! I edited README.md slightly. Could you please resolve the merge conflicts? Then I can merge the PR.

RJ722 commented:
@jendrikseipp, thanks for your timely reviews, and help on this! :-)
